cfg/githubfieldmapper.go

```diff
@@ -0,0 +1,235 @@
+
+	GetFieldValue(jIssue *jira.Issue, fieldKey FieldKey) (interface{}, error)
+
+	// getFieldIDs requests the metadata of every issue field in the JIRA
+	// project, and saves the IDs of the custom fields used by issue-sync.
+	GetFieldIDs(client jira.Client) (map[FieldKey]string, error)
+}
+
+type DefaultFieldMapper struct {
+	Config *Config
+}
+
+type JsonFieldMapper struct {
+	Config *Config
+}
+
+func (m DefaultFieldMapper) GetFieldValue(jIssue *jira.Issue, fieldKey FieldKey) (interface{}, error) {
+	switch fieldKey {
+	case GitHubID:
+		fallthrough
+	case GitHubNumber:
+		return jIssue.Fields.Unknowns.Int(m.Config.GetCompleteFieldKey(fieldKey))
+	default:
+		result, exists := jIssue.Fields.Unknowns.Value(m.Config.GetCompleteFieldKey(fieldKey))
+		if !exists {
+			return nil, errors.New("Field not found")
+		}
+		return result, nil
+	}
+}
+
+func (m DefaultFieldMapper) MapFields(issue *models.ExtendedGithubIssue) jira.IssueFields {
+	fields := jira.IssueFields{
+		Type: jira.IssueType{
+			Name: "Task", // TODO: Determine issue type
+		},
+		Project:     m.Config.GetProject(),
+		Summary:     issue.GetTitle(),
+		Description: issue.GetBody(),
+		Labels: 	 issue.CommitIds,
+		Unknowns:    map[string]interface{}{},
+	}
+
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubID)] = issue.GetID()
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubNumber)] = issue.GetNumber()
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubStatus)] = issue.GetState()
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubReporter)] = issue.User.GetLogin()
+
+	githubLabels := make([]string, len(issue.Labels))
+	for i, v := range issue.Labels {
+		githubLabels[i] = *v.Name
+	}
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubLabels)] = strings.Join(githubLabels, ",")
+
+	fields.Unknowns[m.Config.GetCompleteFieldKey(LastISUpdate)] = time.Now().Format(DateFormat)
+
+	return fields
+}
+
+func (m DefaultFieldMapper) GetFieldIDs(client jira.Client) (map[FieldKey]string, error) {
+	m.Config.log.Debug("Collecting field IDs.")
+	req, err := client.NewRequest("GET", "/rest/api/2/field", nil)
+	if err != nil {
+		return map[FieldKey]string{}, err
+	}
+	jFields := new([]jiraField)
+
+	_, err = client.Do(req, jFields)
+	if err != nil {
+		return map[FieldKey]string{}, err
+	}
+
+	fieldIDs := map[FieldKey]string{}
+
+	for _, field := range *jFields {
+		switch field.Name {
+		case "GitHub ID":
+			fieldIDs[GitHubID] = fmt.Sprint(field.Schema.CustomID)
+		case "GitHub Number":
+			fieldIDs[GitHubNumber] = fmt.Sprint(field.Schema.CustomID)
+		case "GitHub Labels":
+			fieldIDs[GitHubLabels] = fmt.Sprint(field.Schema.CustomID)
+		case "GitHub Status":
+			fieldIDs[GitHubStatus] = fmt.Sprint(field.Schema.CustomID)
+		case "GitHub Reporter":
+			fieldIDs[GitHubReporter] = fmt.Sprint(field.Schema.CustomID)
+		case "Last Issue-Sync Update":
+			fieldIDs[LastISUpdate] = fmt.Sprint(field.Schema.CustomID)
+		}
+	}
+
+	_, ok := fieldIDs[GitHubID]
```

Comment:
I haven't thought it through, but I have a gut feeling that the rest of this method could be made much simpler. (I also don't know if you inherited this from the original code.)

Reply:
I did inherit it from previous code however (without having thought too hard about it) my feeling is this might be as simple as I can make this with the amount of `go` syntax I've learned thus far.